### PR TITLE
fix(datepicker): Close date picker after selecting the same date

### PR DESCRIPTION
**Pull request by chenyuzhcy**
#1783

### CODE REVIEW

File: src/datepicker/datepicker.ts [hunk 1]
```diff
@@ -203,6 +203,11 @@ export class NgbDatepicker implements OnDestroy,
    */
   @Output() navigate = new EventEmitter<NgbDatepickerNavigateEvent>();
 
+  /**
+   * An event fired when select happens.
+   */
+  @Output() onSelect = new EventEmitter<NgbDate>();
```
Comment (maxokorokov) [inline src/datepicker/datepicker.ts]: Could you please change this to: 
```javascript
@Output() select = new EventEmitter<NgbDateStruct>();
```

select → to keep naming consistent
`NgbDateStruct` → as `NgbDate` is internal implementation

File: src/datepicker/datepicker.ts [hunk 2]
```diff
@@ -222,6 +228,12 @@ export class NgbDatepicker implements OnDestroy,
     this.showWeekNumbers = config.showWeekNumbers;
     this.startDate = config.startDate;
 
+    this._dateSelectedSubscription = _service.dateSelected$.subscribe(selected => {
+      if (selected) {
+        this.onSelect.emit();
```
Comment (maxokorokov) [inline src/datepicker/datepicker.ts]: it should emit a `NgbDateStruct`

File: src/datepicker/datepicker.spec.ts [hunk 3]
```diff
@@ -442,6 +442,18 @@ describe('ngb-datepicker', () => {
     expect(document.activeElement).toBe(datepicker.nativeElement);
   });
 
+  it('should emit select event when select date', () => {
```
Comment (maxokorokov) [inline src/datepicker/datepicker.spec.ts]: You should also add the tests for:

- clicking on the same date twice
- pressing Enter key on the same date twice

File: src/datepicker/datepicker-service.ts [hunk 4]
```diff
@@ -12,11 +12,15 @@ import {filter} from 'rxjs/operator/filter';
...
   private _model$ = new Subject<DatepickerViewModel>();
 
+  private _dateSelected$ = new Subject<boolean>();
+
   private _state: DatepickerViewModel =
       {disabled: false, displayMonths: 1, firstDayOfWeek: 1, focusVisible: false, months: [], selectedDate: null};
 
   get model$() { return filter.call(this._model$.asObservable(), model => model.months.length > 0); }
 
+  get dateSelected$() { return filter.call(this._dateSelected$.asObservable(), selected => !!selected); }
```
Comment (maxokorokov) [inline src/datepicker/datepicker-service.ts]: please add the following tests to the `datepicker-service.spec.ts`:

- `focus()` → should emit date regardless if it's the same one or not
- `focusSelect()` → should emit date regardless if it's the same one or not

File: src/datepicker/datepicker-service.ts [hunk 5]
```diff
@@ -87,6 +91,7 @@ export class NgbDatepickerService {
     const validDate = this.toValidDate(date, null);
     if (!this._state.disabled && isChangedDate(this._state.selectedDate, validDate)) {
       this._nextState({selectedDate: validDate});
+      this._dateSelected$.next(true);
```
[the remark this answers is not in the document]
Comment (chenyuzhcy) [reply src/datepicker/datepicker-service.ts]: have no idea how that check got back there, wasn't there when I tested it;-( thanks for pointing that out.

**Comment by maxokorokov** [comment]
Hey, @chenyuzhcy. Thanks a lot (especially for many tests), I've added a couple of changes on top, if you don't mind:

- squashed your commits for clean history
- renamed `onSelect` event to `select` in public `NgdDatepicker` API
- changed `manuallySelect(date)` to `select(date, options: {emitEvent: boolean})` as they usually do it in Angular APIs and I don't think another entry method is necessary in `NgbDatepickerService`
- found a couple of small bugs and added tests for them

Will merge when Travis is green

**Pull request merged**